### PR TITLE
Add funtions to manage users

### PR DESCRIPTION
**Pull request by JanLuca**
Not all changes are fully tested yet...

### CODE REVIEW

File: app/user/templates/admin/deleteUser.html [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{% extends "base.html" %}
+{% import "bootstrap/wtf.html" as wtf %}
+
+{% block title %}Delete user "{{ user.username | safe }}"{% endblock %}
+
+{% block content %}
+  <div class="container">
+    <form class="form-deleteUser" method="POST">
+      {{ wtf.form_errors(form) }}
+      <h4 class="form-deleteUser-heading">Are you sure to delete user "{{ user.username | safe }}"?</h4>
```
Comment (fabianfreyer) [inline app/user/templates/admin/deleteUser.html]: It would be cool if we could abstract this kind of "Are you sure you want to {{action}}" behaviour since we need that a lot

File: app/user/__init__.py [hunk 2]
```diff
@@ -25,21 +43,3 @@ def load_user(id):
     # Set up sanity checks.
     from . import sanity
     getattr(app, 'sanity_check_modules', []).append(sanity)
-
```
Comment (fabianfreyer) [inline app/user/__init__.py]: Is this just code movement? If yes, why, and can we put that in a separate commit?

File: app/user/admin.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+from flask import render_template, redirect, url_for, flash
+from flask_login import login_required
+from flask_wtf import FlaskForm
+from wtforms import SubmitField
+from . import groups_required, user_blueprint
+from .models import User
+
+class DeleteUserForm(FlaskForm):
```
Comment (fabianfreyer) [inline app/user/admin.py]: Again, would be cool to have this as a general yes/no/maybe form.

File: app/user/templates/admin/editUser.html [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+{% endif -%}
+{% endif -%}
+{% endmacro -%}
+
+{% block content %}
+<div class="container">
+  <form class="form-editUser" method="POST">
+    {{ show_form_errors(form) }}
+    <div class="form-group">
+      {{ form.username.label(class='col-sm-2 control-label') | safe }}
```
Comment (fabianfreyer) [inline app/user/templates/admin/editUser.html]: I'm not sure we should assume the label is safe, at least the [reference implementation on wtforms](http://flask.pocoo.org/docs/0.12/patterns/wtforms/#forms-in-templates) doesn't.

File: app/user/templates/admin/listUsers.html [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  <table class='table'>
+    <tr>
+      <th>Username</th>
+      <th>Full name</th>
+      <th>E-Mail</th>
+      <th>Actions</th>
+    </tr>
+    {% for user in users %}
+    <tr>
+      <td>{{ user.username | safe }}</td>
```
Comment (fabianfreyer) [inline app/user/templates/admin/listUsers.html]: these aren't safe.

File: app/user/views.py [hunk 6]
```diff
@@ -116,3 +116,8 @@ def logout():
             return redirect(next)
         else:
             return redirect(url_for("user.home"))
+
+@user_blueprint.route('/editMe', methods=['GET', 'POST'])
```
Comment (fabianfreyer) [inline app/user/views.py]: I think something like `/user/edit` would make more sense here.

File: app/views.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+from wtforms import SubmitField
+
+class ConfirmationForm(FlaskForm):
+    submit = SubmitField('Do it')
+
+class ConfirmationFormData(object):
+    def __init__(self, title=None, action=None, backTo=None):
+        self.title = title
+        self.action = action
+        self.backTo = backTo
```
Comment (JanLuca) [inline app/views.py]: I thought about this but then decide not to write the function for just three lines

File: app/user/admin.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+def listUsers():
+    users = User.query()
+    return render_template('admin/listUsers.html',
+        users = users
+    )
+
+@user_blueprint.route('/admin/user/<string:username>/delete', methods=['GET', 'POST'])
+@login_required
+@groups_required('admin')
+def deleteUser(username):
```
Comment (JanLuca) [inline app/user/admin.py]: Yeah, I can change this, I only like the Java/C++-Style more ;)

File: app/user/templates/admin/editUser.html (hunk 4 above)
Comment (JanLuca) [inline app/user/templates/admin/editUser.html]: The false understanding of {{ ... | safe }}, I will change this. We should make a commit to adjust this everywhere else

File: app/user/views.py (hunk 6 above)
Comment (JanLuca) [inline app/user/views.py]: Sounds good, will be changed.

File: app/user/templates/admin/editUser.html (hunk 4 above)
Comment (fabianfreyer) [inline app/user/templates/admin/editUser.html]: Yup.

File: app/views.py (hunk 7 above)
Comment (fabianfreyer) [inline app/views.py]: Yeah, sounds good. Would probably be useful though to prepare for using the confirm form for confirming deletion of registrations too...

Comment (fabianfreyer) [inline app/views.py]: Yeah, that's a valid design choice. Would probably be useful though to prepare for using the confirm form for confirming deletion of registrations too...

File: app/user/admin.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        flash('User has been deleted', 'success')
+        return redirect(url_for('user.list_users'))
+    return render_template('confirmationPage.html',
+        form = form,
+        data = data
+    )
+
+@user_blueprint.route('/admin/user/<string:username>/edit', methods=['GET', 'POST'])
+@login_required
+def edit_user(username, directCall = True):
```
Comment (fabianfreyer) [inline app/user/admin.py]: What is directCall? Would be good to have a comment / docstr. Also, pep8... direct_call?

File: app/user/admin.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    user = User.get(username)
+    if not user:
+        flash('Invalid user name!', 'error')
+        if directCall:
```
Comment (fabianfreyer) [inline app/user/admin.py]: actually it might make sense to have a [secure redirect back](http://flask.pocoo.org/snippets/62/) thing going on here.

File: app/views.py (hunk 7 above)
Comment (fabianfreyer) [inline app/views.py]: I guess it would be best to merge #34 first, and if you decide to go the decorator way, we could use the decorator here too.

File: app/user/admin.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+def list_users():
+    users = User.query()
+    return render_template('admin/listUsers.html',
+        users = users
+    )
+
+@user_blueprint.route('/admin/user/<string:username>/delete', methods=['GET', 'POST'])
+@login_required
+@groups_required('admin')
+def delete_user(username):
```
Comment (fabianfreyer) [inline app/user/admin.py]: If you make #34 a decorator, this would become way simpler.

**Comment by fabianfreyer** [comment]
I've pushed a [rebased and amended version of this branch](https://github.com/fabianfreyer/account_management/tree/userfuncs), it might make sense to reset to that. Unfortunately it seems I couldn't push directly to this PR.

File: app/user/templates/admin/listUsers.html [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    </tr>
+    {% for user in users %}
+    <tr>
+      <td>{{ user.username }}</td>
+      <td>{{ user.full_name }}</td>
+      <td>{{ user.mail }}</td>
+      <td>
+        <div class="btn-group" role="group">
+          <a class="btn-default btn" href="{{url_for('user.edit_user', username=user.username)}}">
+            <span class="glyphicon glyphicon-edit" aria-hidden="true"></span> Edit
```
Comment (fabianfreyer) [inline app/user/templates/admin/listUsers.html]: `&nbsp` between the `</span>` and `Edit`

File: app/user/admin.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+from app.views import ConfirmationForm, ConfirmationFormData
+from . import groups_required, user_blueprint
+from .models import User
+
+class UserEditForm(FlaskForm):
+    username = StringField('Username')
+    givenName = StringField('Given Name', validators=[DataRequired('Please enter a given name')])
+    surname = StringField('Surname', validators=[DataRequired('Please enter a surname')])
+    mail = EmailField('E-Mail', validators=[
+            DataRequired('Please enter a E-Mail address'),
```
Comment (fabianfreyer) [inline app/user/admin.py]: an E-Mail address

File: app/user/admin.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+        flash('Invalid user name!', 'error')
+        return redirect(url_for('user.list_users'))
+    user.delete()
+    flash('User has been deleted', 'success')
+    return redirect(url_for('user.list_users'))
+
+@user_blueprint.route('/admin/user/<string:username>/edit', methods=['GET', 'POST'])
+@login_required
+def edit_user(username, back_url = None):
+    if not back_url or not is_safe_url(back_url):
```
Comment (fabianfreyer) [inline app/user/admin.py]: minor nitpick here: applying De Morgan's theorem once would lead to a more readable statement:
```python
if not (back_url and is_safe_url(back_url)):
```

Actually it might make sense to just make is_safe_url return `False` in this case since `None` is obviously not a safe URL.
